File: 0207-course-schedule/0207-course-schedule.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        vector<vector<int>> adj(numCourses);
		for (auto it : prerequisites) {
			adj[it[1]].push_back(it[0]);
		}

        vector<int> indeg(numCourses, 0);
        for(int i=0;i<adj.size();i++){
            for(auto it:adj[i]){
                indeg[it]++;
            }
        }

        queue<int>q;
        for(int i=0;i<numCourses;i++){
            if(indeg[i]==0){
                q.push(i);
            }
        }
        int cnt=0;
        while(!q.empty()){
            auto top=q.front();
            q.pop();
            cnt++;
            for(auto it:adj[top]){
                indeg[it]--;
                if(indeg[it]==0){
                    q.push(it);
                }
            }
        }

        if(cnt==numCourses){
            return true;
        }
        return false;
    }
};

File: 0207-course-schedule/0207-course-schedule.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        List<List<Integer>> adj=new ArrayList<>();
        for(int i=0;i<numCourses;i++){
            adj.add(new ArrayList<>());
        }

        int[] indeg=new int[numCourses];

        for(int[] pre:prerequisites){
            int c=pre[0];
            int r=pre[1];
            adj.get(r).add(c);
            indeg[c]++;
        }

        Queue<Integer>q=new LinkedList<>();

        for(int i=0;i<numCourses;i++){
            if(indeg[i]==0){
                q.offer(i);
            }
        }

        int taken=0;

        while(!q.isEmpty()){
            int curr=q.poll();
            taken++;

            for(int next : adj.get(curr)){
                indeg[next]--;
                if(indeg[next]==0){
                    q.offer(next);
                }
            }
        }

        return taken==numCourses;

    }
}